### PR TITLE
kernel: extract ksud to /dev

### PR DESCRIPTION
**Pull request by Ylarod**

### CODE REVIEW

File: kernel/ksud.c [hunk 1]
```diff
@@ -15,24 +17,26 @@
 #include "klog.h" // IWYU pragma: keep
 #include "selinux/selinux.h"
 
-static const char KERNEL_SU_RC[] =
+static char LEGACY_KSUD_PATH[] = "/data/adb/ksud";
```
Comment (tiann) [inline kernel/ksud.c:20]: we don't need backward compatibility, delete this

File: kernel/ksu.c [hunk 2]
```diff
@@ -53,10 +53,10 @@ int __init kernelsu_init(void)
 
 	ksu_uid_observer_init();
 
-#ifdef CONFIG_KPROBES
```
Comment (tiann) [inline kernel/ksu.c:56]: Do we need to delete this?

File: kernel/ksud.c [hunk 3]
```diff
@@ -60,6 +64,72 @@ void on_post_fs_data(void)
 	ksu_load_allow_list();
 }
 
+static struct work_struct extract_ksud_work;
+extern unsigned int ksud_size;
+extern const char ksud[];
```
Comment (tiann) [inline kernel/ksud.c:69]: prefix it with something like `__` or rename it to avoid conflict

File: kernel/ksud.c [hunk 4]
```diff
@@ -60,6 +64,72 @@ void on_post_fs_data(void)
...
+			pr_info("extract_ksud open failed, error: %d",
+				PTR_ERR(fp));
+			return;
+		}
+		pr_info("random ksud: %s", ksu_random_path);
+		snprintf(KERNEL_SU_RC, 1024, KERNEL_SU_RC_TEMPLATE, ksu_random_path,
+			 ksu_random_path, ksu_random_path, ksu_random_path);
+		pr_info("KERNEL_SU_RC:");
+		pr_info("%s", KERNEL_SU_RC);
+		kernel_write(fp, ksud, ksud_size, NULL);
```
Comment (tiann) [inline kernel/ksud.c:118]: print the result and check if it succeeds

File: kernel/ksud.c [hunk 5]
```diff
@@ -260,6 +335,11 @@ static void stop_execve_hook()
 // ksud: module support
 void ksu_enable_ksud()
 {
+	// use legacy path for fallback
+	snprintf(KERNEL_SU_RC, 1024, KERNEL_SU_RC_TEMPLATE, LEGACY_KSUD_PATH,
```
Comment (tiann) [inline kernel/ksud.c:339]: we are alpha version, don't consider compatibility

**Comment by Ylarod** [comment]
```
rubens:/ # dmesg | grep KernelSU
[    0.561086] KernelSU: ksu_lsm_hook_init
[    0.564305] KernelSU: sucompat: execve_kp: 0
[    0.567385] KernelSU: sucompat: newfstatat_kp: 0
[    0.570445] KernelSU: sucompat: faccessat_kp: 0
[    0.570450] KernelSU: KERNEL_SU_RC:
[    0.570452] KernelSU: \x0aon post-fs-data\x0a    exec u:r:su:s0 root -- /data/adb/ksud post-fs-data\x0a\x0aon nonencrypted\x0a    exec u:r:su:s0 root -- /data/adb/ksud services\x0a\x0aon property:vold.decrypt=trigger_restart_framework\x0a    exec u:r:su:s0 root -- /data/adb/ksud services\x0a\x0aon property:sys.boot_completed=1\x0a    exec u:r:su:s0 root -- /data/adb/ksud boot-completed\x0a\x0a
[    0.573476] KernelSU: ksud: execve_kp: 0
[    0.576527] KernelSU: ksud: vfs_read_kp: 0
[    1.813104] KernelSU: extract_ksud kprobe: 1!
[    1.813196] KernelSU: extract_ksud
[    1.813223] KernelSU: random ksud: /dev/ksud_875718d61af21b41ee92169dac2c2dbb
[    1.813228] KernelSU: KERNEL_SU_RC:
[    1.813232] KernelSU: \x0aon post-fs-data\x0a    exec u:r:su:s0 root -- /dev/ksud_875718d61af21b41ee92169dac2c2dbb post-fs-data\x0a\x0aon nonencrypted\x0a    exec u:r:su:s0 root -- /dev/ksud_875718d61af21b41ee92169dac2c2dbb services\x0a\x0aon property:vold.decrypt=trigger_restart_framework\x0a    exec u:r:su:s0 root -- /dev/ksud_875718d61af21b41ee92169dac2c2dbb services\x0a\x0aon property:sys.boot_completed=1\x0a    exec u:r:su:s0 root -- /dev/ksud_875718d61af21b41ee92169dac2c2dbb boot-completed\x0a\x0a
[    1.972321] KernelSU: /system/bin/init second_stage executed
[    2.063759] KernelSU: vfs_read: /system/etc/init/atrace.rc, comm: init, count: 1024, rc_count: 448
[    2.063908] KernelSU: unregister vfs_read kprobe: 1!
[    5.220272] KernelSU: option: 0xdeadbeef, cmd: 7
[    5.220291] KernelSU: post-fs-data triggered
[    5.220293] KernelSU: ksu_load_allow_list
[    5.221913] KernelSU: allowlist version: 1
[    5.221932] KernelSU: load_allow_uid: 2000, allow: 1
[    5.221940] KernelSU: load_allow_uid: 10268, allow: 1
[    5.221981] KernelSU: load_allow_uid: 10299, allow: 1
[    5.221989] KernelSU: load_allow_uid: 10234, allow: 1
[    5.221996] KernelSU: load_allow_uid: 10374, allow: 1
[    5.222003] KernelSU: load_allow_uid: 10420, allow: 0
[    5.222009] KernelSU: load_allow_uid: 10287, allow: 0
[    5.222015] KernelSU: load_allow_list read err: 0
[    5.222019] KernelSU: ksu_show_allow_list
[    5.222021] KernelSU: uid :2000, allow: 1
[    5.222028] KernelSU: uid :10268, allow: 1
[    5.222032] KernelSU: uid :10299, allow: 1
[    5.222036] KernelSU: uid :10234, allow: 1
[    5.222040] KernelSU: uid :10374, allow: 1
[    5.222044] KernelSU: uid :10420, allow: 0
[    5.222048] KernelSU: uid :10287, allow: 0
[    5.634716] KernelSU: exec app_process, /data prepared!
[    5.634725] KernelSU: on_post_fs_data already done
[    5.634729] KernelSU: unregister execve kprobe: 1!
[   12.707125] KernelSU: renameat: packages.list.tmp -> packages.list, new path: /system/packages.list
[   19.420434] KernelSU: renameat: packages.list.tmp -> packages.list, new path: /system/packages.list
[   19.674258] KernelSU: option: 0xdeadbeef, cmd: 7
[   19.674263] KernelSU: boot_complete triggered
[   26.035485] KernelSU: renameat: packages.list.tmp -> packages.list, new path: /system/packages.list
[   26.127410] KernelSU: renameat: packages.list.tmp -> packages.list, new path: /system/packages.list
[   36.035817] KernelSU: renameat: packages.list.tmp -> packages.list, new path: /system/packages.list
[   77.331784] KernelSU: newfstatat su->sh!
[   77.331854] KernelSU: faccessat su->sh!
[   77.332948] KernelSU: do_execveat_common su found
[   77.332985] KernelSU: error: 0, sid: 1171
[   83.954755] KernelSU: renameat: packages.list.tmp -> packages.list, new path: /system/packages.list
```

File: kernel/ksud.c [hunk 6]
```diff
@@ -60,6 +64,72 @@ void on_post_fs_data(void)
...
+// 					 NULL);
+// 	info->path = path;
+// 	if (info == NULL)
+// 		return -ENOMEM;
+// 	return call_usermodehelper_exec(info, wait);
+// }
+
+static void do_extract_ksud(struct work_struct *work)
+{
+	if (ksud_size > 0) {
```
Comment (tiann) [inline kernel/ksud.c:100]: `if (ksud_size <=0) return;`

File: kernel/ksud.c [hunk 7]
```diff
@@ -60,6 +64,72 @@ void on_post_fs_data(void)
...
+	static char *hex = "0123456789abcdef";
+	static char byte;
+	int i;
+	for (i = 0; i < len; i++) {
+		get_random_bytes(&byte, 1);
+		buf[i] = hex[byte % 16];
+	}
+}
+
+// static int call_usermod(const char *path, char **argv, char **envp, int wait)
```
Comment (tiann) [inline kernel/ksud.c:86]: delete it

**Review by tiann — CHANGES_REQUESTED** [review]

File: kernel/ksu.c (hunk 2 above)
Comment (Ylarod) [reply kernel/ksu.c:56]: already checked in ksu_enable_ksud and ksu_enable_sucompat

**Comment by wxt1221** [comment]
Why close this? @Ylarod

**Pull request closed without merging**